File: 153-find-minimum-in-rotated-sorted-array/find-minimum-in-rotated-sorted-array.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        if len(nums)==2:
          return min(nums)
        res = 5001
        left, right = 0, len(nums)-1
        while(left<=right):
          if nums[left] < nums[right]:
            res = min(res,nums[left])
            break
          
          mid = (left+right)//2
          res = min(nums[mid],res)
          if(nums[mid]>=nums[left]):
            left = mid+1
          else:
            right = mid-1
        return res


        

File: 153-find-minimum-in-rotated-sorted-array/find-minimum-in-rotated-sorted-array.java
class Solution {
    public int findMin(int[] a) {
        int l=0, r=a.length-1;
        while(l <= r){
            int m = l + (r-l)/2;

            if(a[r] > a[l]){
                //normal binary search
                return a[l];
            }else{
                //detect which sub array we are in
                if(m+1 < a.length && a[m] > a[m+1]){
                    return a[m+1];
                }
                
                if(m-1 >= 0 && a[m] < a[m-1]){
                    return a[m];
                }

                //we are in right sub array
                if(a[m] < a[l]){
                    r = m - 1;
                }else{
                    l = m + 1;
                }
            }
        }
        return a[0];
    }
}